pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,793 @@
+			backoff = min(backoff*2, backoffMax)
+			time.Sleep(backoff)
+		}
+	}
+}
+
+func (p *Plugin) refreshAliasesTask(ctx context.Context) {
+	ticker := p.clk.Ticker(refreshAliasesFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping refresh keys task", reasonTag, ctx.Err())
+			return
+		case <-ticker.C:
+			p.refreshAliases(ctx)
+		}
+	}
+}
+
+func (p *Plugin) refreshAliases(ctx context.Context) {
+	p.log.Debug("Refreshing aliases")
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	for _, entry := range p.entries {
+		_, err := p.kmsClient.UpdateAlias(ctx, &kms.UpdateAliasInput{
+			AliasName:   &entry.AliasName,
+			TargetKeyId: &entry.Arn,
+		})
+		if err != nil {
+			p.log.Error("It was not possible to refresh alias",
+				aliasNameTag, entry.AliasName, keyArnTag, entry.Arn, reasonTag, err)
+		}
+	}
+}
+
+func (p *Plugin) disposeAliasesTask(ctx context.Context) {
+	ticker := p.clk.Ticker(disposeAliasesFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping dispose aliases task", reasonTag, ctx.Err())
+			return
+		case <-ticker.C:
+			p.disposeAliases(ctx)
+		}
+	}
+}
+
+func (p *Plugin) disposeAliases(ctx context.Context) {
+	p.log.Debug("Looking for aliases in trust domain to dispose")
+	paginator := kms.NewListAliasesPaginator(p.kmsClient, &kms.ListAliasesInput{Limit: aws.Int32(100)})
+
+	for {
+		aliasesResp, err := paginator.NextPage(ctx)
+		switch {
+		case err != nil:
+			p.log.Error("Failed to fetch aliases to dispose", reasonTag, err)
+			return
+		case aliasesResp == nil:
+			p.log.Error("Failed to fetch aliases to dispose: nil response")
+			return
+		}
+
+		for _, alias := range aliasesResp.Aliases {
+			switch {
+			case alias.AliasName == nil || alias.LastUpdatedDate == nil || alias.AliasArn == nil:
+				continue
+				// if alias does not belong to trust domain skip
+			case !strings.HasPrefix(*alias.AliasName, p.aliasPrefixForTrustDomain()):
+				continue
+			// if alias belongs to current server skip
+			case strings.HasPrefix(*alias.AliasName, p.aliasPrefixForServer()):
+				continue
+			}
+
+			now := time.Now()
+			diff := now.Sub(*alias.LastUpdatedDate)
+			// diff should never be negative, just being extra careful
+			if diff < 0 || diff < aliasThreshold {
+				continue
+			}
+			log := p.log.With(aliasNameTag, alias.AliasName)
+			log.Debug("Found alias in trust domain beyond threshold")
+
+			describeResp, err := p.kmsClient.DescribeKey(ctx, &kms.DescribeKeyInput{KeyId: alias.AliasArn})
+			switch {
+			case err != nil:
+				log.Error("Failed to describe key to dispose", reasonTag, err)
+				continue
+			case describeResp == nil || describeResp.KeyMetadata == nil || describeResp.KeyMetadata.Arn == nil:
+				log.Error("Malformed describe key response while trying to dispose")
+				continue
+			}
+
+			log = log.With(keyArnTag, *describeResp.KeyMetadata.Arn)
+			select {
+			case p.scheduleDelete <- *describeResp.KeyMetadata.Arn:
```

Comment:
Shouldn't we be deleting the alias here?

Reply:
But this made me think about a case, where just one server in the cluster was down for too long. After coming back (and if the keys are still scheduled for deletion in AWS) it wouldn't be able to generate the keys.

Reply:
> Aliases are also released once the key is finally deleted.
 Ah, TIL.